src/service.py

```diff
@@ -49,24 +51,71 @@ def configure(self, snap_config: dict[str, Any]) -> None:
         self.snap_client.set(snap_config, typed=True)
 
 
-def snap_install(resource: str) -> SnapService:
-    """Install or refresh the snap, and return the snap service.
+def snap_install(resource: Optional[str], channel: str) -> SnapService:
+    """Install the snap, and return the snap service.
+
+    Before installing the snap, it will try to remove the upstream snap that could be installed on
+    older versions of this charm.
+
+    If the channel is changed, the snap.add method is able to identify and refresh the charm to the
+    new channel.
 
     Raises an exception on error.
     """
+    remove_upstream_snap()
```

Comment:
**issue (blocking):** 

If the resource is empty, the upstream snap will be also removed.

Reply:
That is intentional. Not sure if it will have people using the snap form the snapstore, but I don't see an issue of removing it independently if it was used as resource or from the store

Reply:
I think this is correct, if the resource is empty we should switch from golang-openstack-exporter to charmed-openstack-exporter